Prevent dust outputs from being created during withdraw from wallet

jmacxx opened the pull request:
This change fixes an issue whereby dust change outputs are inadvertently created when users make withdrawals from their wallets.  (Funds -> Send Funds)

The solution taken here is to detect a dust TXO during the withdrawal fee estimation process and add that amount to the fee thus eliminating the dust output.  

For example if the user has 1 BTC and goes to withdraw 0.99999900 BTC it will detect a change TXO of 100 sats which is below the dust limit, increase the fee by 100 sats and therefore withdraw 1 BTC.

This fix only applies to user withdrawals from their wallet.  Other use cases such as P2P trading, deposits and fees will be handled separately.

Related to #4039

Commit:
Remove dust outputs created during withdraw from wallet

This change fixes an issue whereby dust change outputs are
inadvertently created when users make withdrawals from their
wallets.  (Funds -> Send Funds)

The solution taken here is to detect a dust TXO during the withdrawal
fee estimation process and add that amount to the fee thus eliminating
the dust output.

For example if the user has 1 BTC and goes to withdraw 0.99999900 BTC
it will detect a change TXO of 100 sats which is below the dust limit,
increase the fee by 100 sats and therefore withdraw 1 BTC.

This fix only applies to user withdrawals from their wallet.  Other
use cases such as P2P trading, deposits and fees will be handled
separately.

Related to #4039

sqrrm commented on desktop/src/main/java/bisq/desktop/main/funds/withdrawal/WithdrawalView.java:
```diff
@@ -629,6 +631,17 @@ public void updateItem(final WithdrawalListItem item, boolean empty) {
+    private Coin getDust(Transaction tx) {
+        Coin dust = Coin.ZERO;
+        for (TransactionOutput txo: tx.getOutputs()) {
+            if (txo.getValue().value < 546) {
```
Better use a constant for values like the dust limit.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/funds/withdrawal/WithdrawalView.java:
```diff
@@ -330,8 +331,9 @@ private void onWithdraw() {
-                sendersAmount = feeExcluded ? amountAsCoin.add(fee) : amountAsCoin;
+                Coin dust = getDust(feeEstimationTransaction);
+                Coin fee = feeEstimationTransaction.getFee().add(dust);
+                sendersAmount = feeExcluded ? amountAsCoin.add(fee) : amountAsCoin.add(dust);
```
The added complexity of the dust on top of the fee makes this code really hard to read. It's not transparent to neither the developer nor the user what will happen here. The dust amount is so low that it's unlikely to cause monetary damage, but users expect a certain amount to be sent and when the dust is added it might cause confusion. (If I got it right,the dust is added as fee or sent amount depending on whether feeExcluded is set or not) 

I suggest making this code easier to follow, and adding a notice to the user that dust might be padded to the fee and make sure it's only padding the fee.

sqrrm replied:
@m52go could you have a look at the user facing language. We're trying to avoid confusion and I feel I might be too deep here to have a good judgement of what a normal user will find helpful and what's just adding to extra confusion.

m52go replied:
Seems straightforward to me overall.

2 small things.

* [satoshi](https://github.com/bisq-network/bisq/pull/4093/files#diff-ff484a282dd668b4e76956ad0b761971R352) should be plural. It's already pluralized below in the 'required transaction fee' line

* might be nice to change the 2 "transaction fee" mentions to "mining fee" to be more specific and match the text on the 'send funds' screen

sqrrm commented on desktop/src/main/java/bisq/desktop/main/funds/withdrawal/WithdrawalView.java:
```diff
@@ -629,6 +631,17 @@ public void updateItem(final WithdrawalListItem item, boolean empty) {
+        for (TransactionOutput txo: tx.getOutputs()) {
+            if (txo.getValue().value < 546) {
+                dust = dust.add(txo.getValue());
+                log.info("dust TXO = {}", txo.getValue().toFriendlyString());
```
As an avid log reader, a bit more context will really help understand what's going on here.

sqrrm reviewed (changes requested):
Thanks for improving the dust handling.

Please see inline comments on the code.

Commit:
Address issues raised in review by sqrrm on 2020-03-26

- added a comment describing the `removeDust` method and its effects.
- use more descriptive variable names.
- made the logging more verbose to help log readers.
- use a constant for the dust limit
- add a notice to the user when dust is padded to the fee

ghost commented:
Here are some tests I found useful:

### TEST 1 - sending full wallet balance minus 1 satoshi

-  Setup - you have funds in your wallet, potentially split into more than 1 UTXO
-  Send your complete wallet balance minus 1 satoshi, option "Amount includes fee".
-  Verify that the confirmation dialog displays: Warning about dust (1 sats) added to fees;
Sending = the full amount of your wallet; Recipient will receive full amount of your wallet - fee
-  Click yes to make the transaction
-  Verify that the receiver gets full amount of your wallet - fee
-  Verify that the transaction gets confirmed on the blockchain.

### TEST 2 - avoid dusty change output (amount excludes fee)

- Setup - Set custom fee to 18 sats per byte
- Setup - you have 1 UTXO in wallet = 1.0 BTC
- Send 0.99995501, option "Amount excludes fee"
- Verify that the confirmation dialog displays: Warning about dust (431 sats) added to fees; Sending 1.0 BTC; Recipient will receive 0.99995501 BTC; Fee is 4499 sats
- Click yes to make the transaction
- Verify that the receiver 0.99995501 BTC
- Verify that the transaction gets confirmed on the blockchain.
 
    
### TEST 3 - avoid dusty change output (amount includes fee)

- Setup - Set custom fee to 18 sats per byte
- Setup - you have 1 UTXO in wallet = 1.0 BTC
- Send 0.99999999, option "Amount includes fee"
- Verify that the confirmation dialog displays:Warning about dust (1 sats) added to fees; Sending 1.0 BTC; Recipient will receive 0.99995930 BTC; Fee is 4069 sats
- Click yes to make the transaction
- Verify that the receiver gets 0.99995931 BTC
- Verify that the transaction gets confirmed on the blockchain.
 

### TEST 4 - normal transaction with change output (amount excludes fee)

- Setup - Set custom fee to 18 sats per byte
- Setup - you have 1 UTXO in wallet = 1.0 BTC
- Send 0.99990000, option "Amount excludes fee"
- Verify that the confirmation dialog displays: Sending 0.99994050 BTC; Recipient will receive 0.9999 BTC; Fee is 4050 sats
- Click yes to make the transaction
- Verify that the receiver gets 0.9999 BTC
- Verify that the transaction gets confirmed on the blockchain.


### TEST 5 - normal transaction with change output (amount includes fee)

- Setup - Set custom fee to 18 sats per byte
- Setup - you have 1 UTXO in wallet = 1.0 BTC
- Send 0.99990000, option "Amount includes fee"
- Verify that the confirmation dialog displays: Sending 0.9999 BTC; Recipient will receive 0.99985950 BTC; Fee is 4050 sats
- Click yes to make the transaction
- Verify that the receiver gets 0.99985950 BTC
- Verify that the transaction gets confirmed on the blockchain.

Commit:
Change the constant used to determine what amount is considered dust

Now using `Restrictions.getMinNonDustOutput()` which equates to 546 sats

Commit:
Change wording of the dust informative message per review by m52go

Moved message text into displaystrings.properties

sqrrm reviewed (approved):
ACK

This is a lot easier to read than current master, good work. Tested and works as expected, good guide to testing as well.
